### ISSUE TITLE
Possible Python3 issue with `iotile build`

### ISSUE THREAD

**Issue by ekkizogloy** [iotile/coretools#512]
I got this error on a build with a python3.7 environment while it works on a 2.7 environment.

```NameError: name 'basestring' is not defined:
  File "/Users/leekkizogloy/arch/src/fw/lib_controller/SConstruct", line 4:
    autobuild.autobuild_arm_library(libname)
  File "/Users/leekkizogloy/arch/env/py37/lib/python3.7/site-packages/iotile/build/config/site_scons/autobuild.py", line 62:
    family.for_all_targets(family.tile.short_name, lambda x: arm.build_library(family.tile, libname, x))
  File "/Users/leekkizogloy/arch/env/py37/lib/python3.7/site-packages/iotile/build/build/build.py", line 325:
    func(target)
  File "/Users/leekkizogloy/arch/env/py37/lib/python3.7/site-packages/iotile/build/config/site_scons/autobuild.py", line 62:
    family.for_all_targets(family.tile.short_name, lambda x: arm.build_library(family.tile, libname, x))
  File "/Users/leekkizogloy/arch/env/py37/lib/python3.7/site-packages/iotile/build/config/site_scons/arm.py", line 131:
    library_env = setup_environment(chip)
  File "/Users/leekkizogloy/arch/env/py37/lib/python3.7/site-packages/iotile/build/config/site_scons/arm.py", line 171:
    env['CPPPATH'] = chip.includes()
  File "/Users/leekkizogloy/arch/env/py37/lib/python3.7/site-packages/iotile/build/build/build.py", line 191:
    if isinstance(prop, basestring):```

**Comment by dkarchmer** [comment]
Scons was historically a Python 2.7 only package. Just recently they release 3.0 which finally adds support for Python 3. But don't we need to first migrate to that version before we can even talk about `iotile build` on Python 3?

**Comment by timburke** [comment]
Yes, the `iotile-build` package still needs to be ported to Python 3 and will not work correctly until it is ported and released.  `Scons` is one part of that but there is also just a bunch of python 3 cleanliness that needs to happen like using the `future` module to handle `basestring` and removing all of the explicit `unicode` calls.

**Comment by ekkizogloy** [comment]
Cool... I'll clean as I go.

**Comment by x24ling** [comment]
Is there a document somewhere that has a quick go-to list of things to
check for code comparability?

**Comment by timburke** [comment]
The best reference I've found is here:
http://python-future.org/compatible_idioms.html

The `future` package is what we settled on for managing python 2 and 3 compatibility in a clean way and is already a dependency in all `CoreTools` and `iotile-analytics` packages.

The biggest things that are going to byte us on transitioning `iotile-build` to python 3 are:
- no `dict.iteritems()`, `dict.itervalues`, `dict.iterkeys()`
- when using a `bytes` object named `foo`, in python 3, `foo[i]` is an integer, in python 2, its a 1 character `bytes` object that needs `ord` to see the integer value.

**Comment by mattrunchey** [comment]
> Scons was historically a Python 2.7 only package. Just recently they release 3.0 which finally adds support for Python 3. But don't we need to first migrate to that version before we can even talk about `iotile build` on Python 3?

We are using SCons 3.0.1, which supports py 2.7 and 3.5+, so we should be able to migrate all of our code to support py3 now as well (cleaning as we go). Working on it as well in whatever capacity I can...

**Comment by mattrunchey** [comment]
#523 is being worked on to address this.
